core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -2407,45 +2410,51 @@ private void loadMetadataInternal(RpcContext rpcContext, LockedInodePath inodePa
         loadDirectoryMetadata(rpcContext, inodePath, context);
 
         if (context.getOptions().getLoadDescendantType() != LoadDescendantPType.NONE) {
-          ListOptions listOptions = ListOptions.defaults();
-          if (context.getOptions().getLoadDescendantType() == LoadDescendantPType.ALL) {
-            listOptions.setRecursive(true);
+          UfsStatus[] children;
+          if (statusCache != null
+              && context.getOptions().getLoadDescendantType() == LoadDescendantPType.ALL) {
+            // Avoid re-caching. Makes the assumption the cache was created on the same ufs URI
+            children = statusCache.values().toArray(new UfsStatus[0]);
```

Comment:
Currently, I perform this check here: https://github.com/Alluxio/alluxio/blob/e85c520726318107e7c563f06fceec577daa61a6/core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java#L3422-L3426

Though it might make more sense to put it within the function? What do you think?

Reply:
yeah, let's put it in the function, so people like me won't miss it.

Reply:
updated